src/xrpld/app/tx/detail/VaultWithdraw.cpp

```diff
@@ -192,72 +192,138 @@ VaultWithdraw::doApply()
     // to deposit into it, and this means you are also indefinitely authorized
     // to withdraw from it.
 
-    auto amount = ctx_.tx[sfAmount];
-    auto const asset = vault->at(sfAsset);
+    auto const amount = ctx_.tx[sfAmount];
+    Asset const vaultAsset = vault->at(sfAsset);
     auto const share = MPTIssue(mptIssuanceID);
-    STAmount shares, assets;
-    if (amount.asset() == asset)
+    STAmount sharesRedeemed = {vault->at(sfShareMPTID)}, assetsWithdrawn;
+    try
     {
-        // Fixed assets, variable shares.
-        assets = amount;
-        shares = assetsToSharesWithdraw(vault, sleIssuance, assets);
+        if (amount.asset() == vaultAsset)
+        {
+            // Fixed assets, variable shares.
+            {
+                auto const maybeShares =
+                    assetsToSharesWithdraw(vault, sleIssuance, amount);
+                if (!maybeShares)
+                    return tecINTERNAL;  // LCOV_EXCL_LINE
+                sharesRedeemed = *maybeShares;
+            }
+
+            if (sharesRedeemed == beast::zero)
+                return tecPRECISION_LOSS;
+            auto const maybeAssets =
```

Comment:
Because of the truncation/recalculation, which is done on deposit and recalculation here then is it possible that if I deposit A and get S; if I withdraw A then I'll redeem more than S and likewise, if I withdraw S then I'll get less than A? Another case would be also bad; i.e. I withdraw A and redeem less than S and I withdraw S and get more than A. The same concern applies to `VaultClawback`.

Reply:
If we are given an A (number of assets) then we _always_ (i.e. in `VaultDeposit` `VaultClawback` `VaultWithdraw`) do _roundtrip_ _calculation_ : first from assets to shares, then from such rounded shares to assets again. This guarantees that the number of assets _always_ corresponds with the number of shares, bar rounding of assets (for MPT or XRP) which can go either way. This is similar how DEX works when one side is XRP - which also can go either way due to rounding. If the asset is an IOU then there is no rounding of assets (other than epsilon) which is also similar to how DEX works.